File: codes/go/route/probe.go
package route

import (
	"net/http"

	"github.com/gin-gonic/gin"

	"github.com/nicolas-grevin/green-it/model"
	"github.com/nicolas-grevin/green-it/service"
)

type probeRoute struct {
  svc *service.PostgresService
}

func ProbeRouter(svc *service.PostgresService, r *gin.RouterGroup) {
  pb := probeRoute{
    svc,
  }

  r.GET("/liveness", pb.liveness)
  r.GET("/readiness", pb.readiness)
}

func (pr *probeRoute) check(c *gin.Context) {
  probe := model.NewProbe()
  postgresServiceProb := model.ServiceStatus{
    Status: "ok",
    Message: "Service is up",
  }

  status := http.StatusOK

  if err := pr.svc.Ping(); err != nil {
    status = http.StatusInternalServerError

    postgresServiceProb.Status = "nok"
    postgresServiceProb.Message = err.Error()
  }

  probe.Services["postgres"] = postgresServiceProb

  c.JSON(status, probe)
}

func (pr *probeRoute) liveness(c *gin.Context) {
  pr.check(c)
}

func (pr *probeRoute) readiness(c *gin.Context) {
  pr.check(c)
}

File: codes/deno/route/probe.ts
import { Hono, Context } from "hono";
import { Client } from "postgres";

import { RouteHandler } from "./route.ts";

import { Probe, ServiceStatus, Services } from "../model/probe.ts";

export class ProbeHandler implements RouteHandler {
  private _client: Client;
  private _router: Hono;

  constructor(client: Client) {
    this._client = client;
    this._router = new Hono();
  }

  getRoutes(): Hono {
    this._router.get("/liveness", (c: Context) => this.liveness(c));
    this._router.get("/readiness", (c: Context) => this.readiness(c));

    return this._router;
  }

  private liveness(c: Context) {
    return this.check(c);
  }
  
  private readiness(c: Context) {
    return this.check(c);
  }

  private async check(c: Context) {
    const postgresServiceStatus: ServiceStatus = {
      status: "ok",
      message: "Service is up"
    };

    const services: Services = {
      "postgres": postgresServiceStatus, 
    };

    const probe: Probe = {
      services
    }

    try {    
      await this._client.queryArray`SELECT 1`;
    } catch (e) {
      postgresServiceStatus.status = "nok";
      postgresServiceStatus.message = "Service is down";
    
      return c.json(probe, 500)
    }

    return c.json(probe, 200);
  }
}
